[GLib] Add async versions of download_uri methods

TingPing opened the pull request:
#### 2822ceaf61b67bcbab54ad910ec2a4cc3320181b
<pre>
[GLib] Add async versions of download_uri methods
<a href="https://bugs.webkit.org/show_bug.cgi?id=289625">https://bugs.webkit.org/show_bug.cgi?id=289625</a>

Reviewed by NOBODY (OOPS!).

WebKit wants to move away from synchronous constructors for WKDownload
so we need to expose async versions in our API.

This covers both WebKitWebView and WebKitNetworkSession.

* Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp:
(webkit_network_session_download_uri_finish):
(webkit_network_session_download_uri_async):
* Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
(runAssertionsForDownloadTest):
(testWebViewDownloadURI):
(testWebViewDownloadURIAsync):
(testSessionDownloadURIAsync):
(beforeAll):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2822ceaf61b67bcbab54ad910ec2a4cc3320181b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94967 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14562 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4419 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99995 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45465 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97016 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14851 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22984 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72446 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29741 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97968 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11077 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85747 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52770 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10786 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3469 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44802 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80995 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3566 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102038 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22009 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16071 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81450 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22256 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81763 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80830 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25389 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2788 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15251 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21982 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27104 "") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21639 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25114 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23380 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2685,31 +2685,41 @@ Ref<DownloadProxy> WebsiteDataStore::createDownloadProxy(Ref<API::DownloadClient
+        request.setFirstPartyForCookies(originatingPageURL);
+        request.setIsSameSite(WebCore::areRegistrableDomainsEqual(originatingPageURL, request.url()));
+        if (!request.hasHTTPHeaderField(WebCore::HTTPHeaderName::UserAgent))
+        request.setHTTPUserAgent(originatingPage->userAgentForURL(request.url()));
```
```suggestion
        if (!request.hasHTTPHeaderField(WebCore::HTTPHeaderName::UserAgent))
            request.setHTTPUserAgent(originatingPage->userAgentForURL(request.url()));
```

it's odd the style checker didn't pick this

spenap commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2685,31 +2685,41 @@ Ref<DownloadProxy> WebsiteDataStore::createDownloadProxy(Ref<API::DownloadClient
+        request.setFirstPartyForCookies(URL());
+        request.setIsSameSite(false);
+        if (!request.hasHTTPHeaderField(WebCore::HTTPHeaderName::UserAgent))
+        request.setHTTPUserAgent(WebPageProxy::standardUserAgent());
```
```suggestion
        if (!request.hasHTTPHeaderField(WebCore::HTTPHeaderName::UserAgent))
            request.setHTTPUserAgent(WebPageProxy::standardUserAgent());
```

spenap replied:
beyond the mere formatting suggestions above, not sure if it's cleaner to move the user agent logic outside, something like

```
if (!request.hasHTTPHeaderField(WebCore::HTTPHeaderName::UserAgent))
    request.setHTTPUserAgent(originatingPage ? originatingPage->userAgentForURL(request.url()) : WebPageProxy::standardUserAgent());
```

TingPing replied:
No strong preference. Was just keeping the logic unchanged.

TingPing committed:
[GLib] Add async versions of download_uri methods
https://bugs.webkit.org/show_bug.cgi?id=289625

Reviewed by NOBODY (OOPS!).

WebKit wants to move away from synchronous constructors for WKDownload
so we need to expose async versions in our API.

This covers both WebKitWebView and WebKitNetworkSession.

* Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp:
(webkit_network_session_download_uri_finish):
(webkit_network_session_download_uri_async):
* Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
(runAssertionsForDownloadTest):
(testWebViewDownloadURI):
(testWebViewDownloadURIAsync):
(testSessionDownloadURIAsync):
(beforeAll):

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp:
```diff
@@ -664,4 +664,66 @@ WebKitDownload* webkit_network_session_download_uri(WebKitNetworkSession* sessio
+  * Finish an asynchronous operation started with webkit_network_session_download_uri_async().
+  *
+  * Returns: (transfer full): a new #WebKitDownload representing the download operation.
+  * Since: 2.48
```
```suggestion
  * Since: 2.50
```

Will need to change the Since tags everywhere.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp:
```diff
@@ -664,4 +664,66 @@ WebKitDownload* webkit_network_session_download_uri(WebKitNetworkSession* sessio
 }
 
+/**
+ * webkit_network_session_download_uri_finish:
+  * @session: a #WebKitNetworkSession
```
Watch your indentation here.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp:
```diff
@@ -664,4 +664,66 @@ WebKitDownload* webkit_network_session_download_uri(WebKitNetworkSession* sessio
+            return;
+
+        if (!downloadProxy) {
+            g_task_return_new_error(task.get(), G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to create download");
```
```suggestion
            g_task_return_new_error(task.get(), G_IO_ERROR, G_IO_ERROR_FAILED, _("Failed to create download"));
```

(Because GErrors have a habit of appearing in user interfaces. Epiphany might display this in a toast, for example.)

carlosgcampos replied:
We have an error domain for downloads, so maybe we can add an error for this, instead of using G_IO_ERROR since this is not about IO. The callback is only called with nullptr from the proxy destructor, which means the proxy is destroyed before the download started. So, I think the error would be Failed to start to download, be cause the download has already been created.

carlosgcampos replied:
I wonder if we should even handle this case, Cocoa has an assert

```
    _page->downloadRequest(request, [completionHandler = makeBlockPtr(completionHandler)] (auto* download) {
        if (download)
            completionHandler(wrapper(download));
        else
            ASSERT_NOT_REACHED();
    });
```

achristensen07 replied:
I think I'm hoping to eventually turn that failure case into a callback with a DownloadProxy object that does nothing but it will immediately receive an error in its WKDownloadDelegate.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5004,6 +5004,74 @@ WebKitDownload* webkit_web_view_download_uri(WebKitWebView* webView, const char*
+            return;
+
+        if (!downloadProxy) {
+            g_task_return_new_error(task.get(), G_IO_ERROR, G_IO_ERROR_FAILED, "Failed to create download");
```
```suggestion
            g_task_return_new_error(task.get(), G_IO_ERROR, G_IO_ERROR_FAILED, _("Failed to create download"));
```

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
```diff
@@ -668,27 +668,54 @@ class WebViewDownloadTest: public WebViewTest {
+#if ENABLE(2022_GLIB_API)
+static void testSessionDownloadURIAsync(WebViewDownloadTest* test, gconstpointer)
+{
+    webkit_network_session_download_uri_async(test->m_networkSession.get(), kServer->getURIForPath("/test.pdf").data(), nullptr, (GAsyncReadyCallback)(+[](WebKitNetworkSession* networkSession, GAsyncResult* result, WebViewDownloadTest* test) {
+        GRefPtr<WebKitDownload> download = adoptGRef(webkit_network_session_download_uri_finish(networkSession, result, nullptr));
+        runAssertionsForDownloadTest(test, download.get());
+    }), test);
+    test->waitUntilDownloadStarted();
```
I assume you've checked to ensure the callbacks in the tests are actually executed? I'm a little surprised this works, since the test only waits until download started, and I would expect the callbacks are not executed until the download finishes.

Even if it works today, I wonder if it's still guaranteed to work tomorrow. It's easy to guard against this, though: just add a `bool testFinished = false` variable, set it to true in the callback, and use `g_assert_true()` after `waitUntilDownloadStarted()`, then we'll know if the test breaks in the future.

mcatanzaro reviewed (approved):
Our rule is that new APIs require approval from two WPE/GTK reviewers. You do get to provide your own review for the purposes of this rule, though, so technically you could land this now. Still, probably best to wait for Carlos or Adrian before landing; the more eyes, the better.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp:
```diff
@@ -664,4 +664,66 @@ WebKitDownload* webkit_network_session_download_uri(WebKitNetworkSession* sessio
+            return;
+        }
+
+        auto download = webkitDownloadCreate(*downloadProxy, nullptr);
```
```suggestion
        auto download = webkitDownloadCreate(*downloadProxy);
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitNetworkSession.cpp:
```diff
@@ -664,4 +664,66 @@ WebKitDownload* webkit_network_session_download_uri(WebKitNetworkSession* sessio
+
+        auto download = webkitDownloadCreate(*downloadProxy, nullptr);
+
+        webkitDownloadStarted(download.get());
+        webkitNetworkSessionDownloadStarted(WEBKIT_NETWORK_SESSION(g_task_get_source_object(task.get())), download.get());
+
+        g_task_return_pointer(task.get(), download.leakRef(), g_object_unref);
```
This is problematic, it's the only case in which we receive a WebKitDownload after download-started has been emitted in the network session. For example, epiphany creates EphyDownload wrapper for WebKitDownload, and assumes that if download-started is emitted and the download is not known, it needs to create the wrapper.

carlosgcampos replied:
I'm not sure I understand the request to make the apis async to be honest, because at this moment the download proxy is always created synchronously. WebKitDownload wraps a DownloadProxy and we already have the download-started signal to notify when it actually starts. If we want to change the concept of WebKitDownload as an already started download then we should not emit download-started for the async apis.

mcatanzaro reviewed (commented):
.

mcatanzaro reviewed (changes requested):
.
